### PR TITLE
Vulkan setup branch

### PR DESCRIPTION
**Pull request by Carlbaum**
Currently not baked into the engine wrappers, just a separate part - integration will be a separate PR.

Please test The helloworld_renderer.cpp when pulling this.

### CODE REVIEW

File: scripts/vulkan_init/README.txt [hunk 1]
```diff
@@ -0,0 +1,43 @@
+VULKAN VALIDATION LAYER SETUP
+
+ON WINDOWS:
```
Comment (novalain) [inline scripts/vulkan_init/README.txt]: Why do we need manual steps and why does the user have to place some .txt file manually? Can't we specify the validation layers somewhere else (in code for example for the time being)?

File: scripts/vulkan_init/vk_layer_settings.txt [hunk 2]
```diff
@@ -0,0 +1,170 @@
+################################################################################
```
Comment (novalain) [inline scripts/vulkan_init/vk_layer_settings.txt]: Is this file really needed? What's it used for? I thought validation layers was part of the SDK

File: scripts/vulkan_init/init_vulkan_windows.bat [hunk 3]
```diff
@@ -0,0 +1 @@
+setx VK_INSTANCE_LAYERS VK_LAYER_LUNARG_standard_validation
```
Comment (novalain) [inline scripts/vulkan_init/init_vulkan_windows.bat]: A bat script? I don't think we want multiple build systems.. Why can't we configure everything in CMake?

File: src/direwolf/src/vulkan/ext/vulkan.h [hunk 4]
```diff
@@ -0,0 +1,79 @@
+#ifndef VULKAN_H_
```
Comment (novalain) [inline src/direwolf/src/vulkan/ext/vulkan.h]: I wouldn't add any vulkan headers manually to the repository. Can't we just use `find_package(Vulkan REQUIRED)` in CMake? And the headers would be `${Vulkan_INCLUDE_DIRS}` ?

File: src/examples/helloworld.cpp [hunk 5]
```diff
@@ -15,5 +15,6 @@ int main() {
 	page::Engine engine(std::move(data));
 	engine.Run();
 
+    std::cout << "\n\nPress enter to EXIT";
```
Comment (novalain) [inline src/examples/helloworld.cpp]: Convert file to spaces :)

File: src/direwolf/src/vulkan/vulkanfunctions.h [hunk 6]
```diff
@@ -0,0 +1,16 @@
+#pragma once
+
+#include "ext/vulkan.h"
+
+namespace page::vulkan {
```
Comment (novalain) [inline src/direwolf/src/vulkan/vulkanfunctions.h]: namespace name

File: src/direwolf/src/vulkan/listofvulkanfunctions.inl [hunk 7]
```diff
@@ -0,0 +1,53 @@
+#ifndef EXPORTED_VULKAN_FUNCTION
```
Comment (novalain) [inline src/direwolf/src/vulkan/listofvulkanfunctions.inl]: What's this file for? Is it needed?

File: scripts/vulkan_init/init_vulkan_windows.bat (hunk 3 above)
Comment (Carlbaum) [reply scripts/vulkan_init/init_vulkan_windows.bat]: It's not for building the project, it's just for enabling the validation layers. 
You probably won't want them enabled at all times as they significantly impact performance, which is why i don't think we should enable it with CMAKE.
So I'd actually say we should have one more script to easily disable them as well ;)

File: src/direwolf/src/vulkan/ext/vulkan.h (hunk 4 above)
Comment (Carlbaum) [reply src/direwolf/src/vulkan/ext/vulkan.h]: You can check the `Preparing for loading Vulkan API functions` section of the vulkan cookbook, it's really well explained why we'd want to do it like this.
Essentially it boils down to saving performance, and together with the `vulkanfunctions.*` and `listofvulkanfunctions.inl` getting more control over what vulkan functionalities that make it into our library.

File: src/examples/helloworld.cpp (hunk 5 above)
Comment (Carlbaum) [reply src/examples/helloworld.cpp]: hahah, sure. So we decided on spaces over tabs now? ;p

File: src/direwolf/src/vulkan/vulkanfunctions.h (hunk 6 above)
Comment (Carlbaum) [reply src/direwolf/src/vulkan/vulkanfunctions.h]: What namespace name have we decided upon then? ;)

File: src/direwolf/src/vulkan/listofvulkanfunctions.inl (hunk 7 above)
Comment (Carlbaum) [reply src/direwolf/src/vulkan/listofvulkanfunctions.inl]: So these 3 `vulkanfunction` files are used for specifying which functions we'll be using within our software. It will require more work than just including everything, but as mentioned in a previous comment - it will give us some increased performance.

File: scripts/vulkan_init/init_vulkan_windows.bat (hunk 3 above)
Comment (novalain) [reply scripts/vulkan_init/init_vulkan_windows.bat]: Sure, I get that they might change often and shouldn't be hardcoded. But I'm very skeptical to bring in platform specific scripts in this project :P I understand that it could be part of the config, but can we make a generic JSON config file or equivalent maybe? Where we specify the validation layers etc

Comment (Carlbaum) [reply scripts/vulkan_init/init_vulkan_windows.bat]: Ah, yeah that makes sense. So how about two platform independent python scripts?
So no matter which platform you're on you just run the `enableVulkanValidationLayers.py` or `disableVulkanValidationLayers.py` from the scripts folder?

File: src/direwolf/src/vulkan/ext/vulkan.h (hunk 4 above)
Comment (novalain) [reply src/direwolf/src/vulkan/ext/vulkan.h]: Okay but looking at those files they aren't part of the `ext` but rather some config. My question is why do we add headers to `ext`? Can we let cmake find vulkan and link against the headers and the library instead?

File: scripts/vulkan_init/init_vulkan_windows.bat (hunk 3 above)
Comment (novalain) [reply scripts/vulkan_init/init_vulkan_windows.bat]: Okay so this is essentially only setting an environment variable. I still think adding external scripts is s sign of a bad project setup.. and should really be avoided if possible.

Can vulkan only read the validation layers from this environment variable? Is this really the only way? I'd prefer to have a JSON config file for vulkan or equivalent (that specifies all the validation layers) that gets parsed automatically as part by the build process and linked against.

File: src/direwolf/src/vulkan/vulkanfunctions.h (hunk 6 above)
Comment (novalain) [reply src/direwolf/src/vulkan/vulkanfunctions.h]: Well. dwolf ? :P :d

File: scripts/vulkan_init/init_vulkan_windows.bat (hunk 3 above)
Comment (novalain) [reply scripts/vulkan_init/init_vulkan_windows.bat]: // Vulkan config.json
```
"Validation Layers" : [
VK_LAYER_LUNARG_standard_validation, 
VK_LAYER_LUNARG_otherlayer1.
VK_LAYER_LUNARG_otherlayer2,
VK_LAYER_LUNARG_otherlayer3,
VK_LAYER_LUNARG_otherlayer4];

```
etc. I'm not saying we should bring in JSON now, but that we need to think about how we should do this.

File: scripts/vulkan_init/README.txt (hunk 1 above)
Comment (novalain) [reply scripts/vulkan_init/README.txt]: Okay It seems weird and hardcoded that it has to be next to the executable. Can we put it in a vulkan specific config folder and copy it next to the bin as part of the build step?

**Comment by novalain** [comment]
Next step move renderer into submodule and `page::vulkan` becomes `dw::vulkan`
